src/object/obj_class.c

```diff
@@ -482,3 +496,76 @@ obj_ec_codec_get(daos_oclass_id_t oc_id)
 
 	return NULL;
 }
+
+
+/**
+ * Encode (using ISA-L) a full stripe from the submitted scatter-gather list.
+ *
+ * oid		[IN]		The object id of the object undergoing encode.
+ * sgl		[IN]		The SGL containing the user data.
+ * sg_idx	[IN|OUT]	Index of sg_iov entry in array.
+ * sg_off	[IN|OUT]	Offset into sg_iovs' io_buf.
+ * parity	[IN|OUT]	Struct containing parity buffers.
+ * p_idx	[IN]		Index into parity p_bufs array.
+ */
+int
+obj_encode_full_stripe(daos_obj_id_t oid, daos_sg_list_t *sgl, uint32_t *sg_idx,
+		       size_t *sg_off, struct obj_ec_parity *parity, int p_idx)
+{
+	struct obj_ec_codec		*codec = obj_ec_codec_get(
+							daos_obj_id2class(oid));
+	struct daos_oclass_attr		*oca = daos_oclass_attr_find(oid);
+	unsigned int			 len = oca->u.ec.e_len;
+	unsigned int			 k = oca->u.ec.e_k;
+	unsigned int			 p = oca->u.ec.e_p;
+	unsigned char			*data[k];
+	unsigned char			*ldata[k];
+	int				 i, lcnt = 0;
+	int				 rc = 0;
+
+	for (i = 0; i < k; i++)
+		if (sgl->sg_iovs[*sg_idx].iov_len - *sg_off >= len) {
+			unsigned char *from =
+				(unsigned char *)sgl->sg_iovs[*sg_idx].iov_buf;
+
+			data[i] = &from[*sg_off];
+			*sg_off += len;
+			if (*sg_off == sgl->sg_iovs[*sg_idx].iov_len) {
+				*sg_off = 0;
+				(*sg_idx)++;
+			}
+		} else {
+			int cp_cnt = 0;
+
+			D_ALLOC(ldata[lcnt], len);
+			if (ldata[lcnt] == NULL)
+				D_GOTO(out, rc = -DER_NOMEM);
+			while (cp_cnt < len) {
+				int cp_amt =
+					sgl->sg_iovs[*sg_idx].iov_len-*sg_off <
+					len - cp_cnt ?
+					sgl->sg_iovs[*sg_idx].iov_len-*sg_off :
+					len - cp_cnt;
+				unsigned char *from =
+					sgl->sg_iovs[*sg_idx].iov_buf;
+
+				memcpy(&ldata[lcnt][cp_cnt], &from[*sg_off],
+				       cp_amt);
+				if (sgl->sg_iovs[*sg_idx].iov_len - *sg_off <
```

Comment:
above "<" should be "<=“？

Reply:
this one can be fixed by following patch.

Reply:
maybe fine to combine the 10 commits in this PR into one. And fix this issue as well, or am I miss-understood it?